pkg: add support for the tinycrypt library

haukepetersen opened the pull request:
This PR adds [tinycrypt](https://github.com/01org/tinycrypt) as a package to RIOT. My main motivation for adding this library is that this is the crypto lib used by the NimBLE bluetooth stack which I am porting at the moment.

Building this package is very simple: all it needs is a trivial Makefile in the source folder, triggered by the packages main Makefile. Both are applied to the tinycrypt package via patch.

A second path fixed a minor style issue with the original code - I PRed the same fix against the tinycrypt repository so we hopefully can omit that path soon.

The test application is (more or less randomly) copied from the `tinycrypt` project, as the `aes` test shows well if the integration was successful or not.

aabadie commented on tests/pkg_tinycrypt/tests/01-run.py:
```diff
@@ -0,0 +1,19 @@
+
+import os
+import sys
+
```
Needs 2 blank lines here ;)

aabadie commented on tests/pkg_tinycrypt/tests/01-run.py:
```diff
@@ -0,0 +1,19 @@
+def testfunc(child):
+    child.expect_exact('PASS - main.')
+    child.expect_exact('PROJECT EXECUTION SUCCESSFUL')
+
```
here as well

aabadie commented on tests/pkg_tinycrypt/test_utils.h:
```diff
@@ -0,0 +1,125 @@
+/*  test_utils.h - TinyCrypt interface to common functions for tests */
```
Even if imported as is, this file should be adapted to RIOT's code standards. Otherwise the static tests fail.

kaspar030 replied:
If it is used as is, maybe adding the package's folder to the include path saves us from dealing with this?

kaspar030 commented on pkg/tinycrypt/patches/0001-adapt-Makefiles.patch:
```diff
@@ -0,0 +1,51 @@
+From a0e94f07ccf9c90ca49a905adc925b4ebfd8f0eb Mon Sep 17 00:00:00 2001
```
I propose copying this file, overwriting the one in the package sources, as other packages (like minmea) do.
Maintaining the patch is way more overhead.

kaspar030 commented on tests/pkg_tinycrypt/main.c:
```diff
@@ -0,0 +1,2078 @@
+/* test_aes.c - TinyCrypt AES-128 tests (including NIST tests) */
```
same here - as the sources are already extracted to the package source folder in bin/pkg, maybe, with the right makefile (and maybe a small patch), it can be reused from there?

haukepetersen replied:
Yes, most likely. I think I just have an idea, let me try something...

haukepetersen replied:
hm, that did not work. Any idea how we could get this done, possibly without patches? Alternatively I tend to simply write a very small test app myselfe, in the end it only needs to to a small number of function calls to the tinycrypt lib, right?

haukepetersen commented:
fixed the test application: python script now pep8 compliant?! and also using the `test_utils.h` header now directly from the tinycrypt sources (no style fixes needed....).

haukepetersen commented:
hm, but still need to look into the patches

haukepetersen commented:
Ok, the style fix in the second patch has already been approved but is waiting for merge in the `tinycrypt` master repository. So I will wait for this fix to be upstream and then we can remove the patches in this PR all together... Stay tuned :-)

haukepetersen commented:
for reference: https://github.com/01org/tinycrypt/pull/29

haukepetersen commented:
alright, the upstream fix to tinycrypt has been merged, so this makes things easier for us: removed both patches and simply copy over the RIOT specific Makefile now. Also remove the trampline for the Makefile in the packages base folder and call the Makefile in the `lib/source` folder now directly. Looks much nicer now, thanks for this hints!

haukepetersen commented:
Also rewrote the test application, now there should not be any trouble with style or code duplication.

haukepetersen commented:
Next problem; tinycrypt does not build nicely for 8- and 16-bit. But we have currently no means of black/whitelisting certain architectures for modules, right?

haukepetersen commented:
but if I see it correctly,  Murdock won't be happy anytime soon...

haukepetersen commented:
rebased and fixed comments.

Regarding the '32-bit only' problem: I whitelisted the test for the `nrf52dk` only for now, as this is the board where the lib is needed first/next (target platform for our initial nimBle porting).

haukepetersen commented:
@kaspar030 @aabadie ping?!

bergzand commented:
> Regarding the '32-bit only' problem: I whitelisted the test for the nrf52dk only for now, as this is the board where the lib is needed first/next (target platform for our initial nimBle porting).

Would it be possible to blacklist all the non-32 bit platforms? I had the same issue with tinycbor and that's what I did there. I'm planning on whitelisting the 32-bit architectures as soon as we have a feature for that.

kaspar030 commented:
> Would it be possible to blacklist all the non-32 bit platforms?

https://github.com/RIOT-OS/RIOT/pull/9081

bergzand commented on tests/pkg_tinycrypt/tests/01-run.py:
```diff
@@ -0,0 +1,20 @@
+
+
+if __name__ == "__main__":
+    sys.path.append(os.path.join(os.environ['RIOTBASE'], 'dist/tools/testrunner'))
```
Could you please use the new `RIOTTOOLS` variable here

haukepetersen commented:
@bergzand addressed pointed out issues

about the 32-bit platform whitelisting: I propose to leave this PR as is for now and update the Makefile as soon as #9081 is merged.

haukepetersen commented:
For the sake of simple testing I also whitelisted native. Should be all good now?!

bergzand commented:
> about the 32-bit platform whitelisting: I propose to leave this PR as is for now and update the Makefile as soon as #9081 is merged.

Ok, fine with me. @kaspar030, @aabadie any further remarks that block squashing and merging this?

haukepetersen commented:
thanks. @aabadie @kaspar030: what do you say?

aabadie commented:
I think you can squash and let @kaspar030 press the green button.

kaspar030 commented on tests/pkg_tinycrypt/Makefile:
```diff
@@ -0,0 +1,13 @@
+
+USEPKG += tinycrypt
+USEMODULE = fmt
+
```
feel free to add ```TEST_ON_CI_WHITELIST += all``` here so the test can be run on CI.

haukepetersen commented:
added `CI_WHITELIST all` and squashed.
